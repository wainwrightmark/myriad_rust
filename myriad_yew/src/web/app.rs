use crate::state::preferences_state::DarkModeState;
use crate::state::prelude::*;
use crate::web::found_words::*;
use crate::web::prelude::*;
use wasm_bindgen_futures::spawn_local;
use yew::prelude::*;
use yew_hooks::use_effect_once;
use yew_router::prelude::*;
use yewdux::prelude::use_store;
use yewdux::prelude::Dispatch;

#[derive(Clone, Routable, PartialEq)]
pub enum Route {
    #[at("/")]
    #[not_found]
    Home,
    #[at("/game/:game")]
    Game { game: String },

    #[at("/cheat/:game")]
    Cheat { game: String },
}

#[function_component(App)]
pub fn app() -> Html {
    //Load the dark mode state here to make sure dark mode is set correctly
    let _dms = use_store::<DarkModeState>();

    use_effect_once(|| {
        spawn_local(crate::web::startup::setup());
        || ()
    });

    html! {
        <BrowserRouter>
            <Switch<Route> render={switch} />
        </BrowserRouter>
    }
}

fn switch(route: Route) -> Html {
    match route {
        Route::Home => {
            let game = Dispatch::<FullGameState>::new()
                .get()
                .game
                .board
                .canonical_string();
            html! { <Redirect<Route> to={Route::Game { game  }} /> }
        }
        Route::Game { game } => {
            let cheat = false;
            html! { <MyriadApp {game} {cheat} />}
        }

        Route::Cheat { game } => {
            let cheat = true;
            html! { <MyriadApp {game} {cheat} />}
        }
    }
}

#[derive(Debug, Properties, PartialEq, Clone)]
struct MyriadAppProps {
    pub game: String,
    pub cheat: bool,
}

#[function_component(MyriadApp)]
fn myriad_app(props: &MyriadAppProps) -> Html {
    let cheat = props.cheat;
    let game = props.game.replace(' ', "+");
    let game = Game::from_string(game.as_str());
    if let Some(game) = game {
        Dispatch::new().apply(LoadGameMessage { game });
    }

    let (width, height) = yew_hooks::use_window_size();
    let (width, height) = (width as f32, height as f32);

    // let node = use_node_ref();
    // let (width, height) = yew_hooks::use_size(node.clone());


    Dispatch::<GameSize>::new().apply(SetSizeMessage { width, height });

    if width == 0. && height == 0. {
        return html!(
            <div class="outer-container">
            <div class="container" />
            </div>
        );
    }

    let size = GameSize::from_width_and_height(width, height);

    let outer_container_style = size.outer_container_style();
    let container_style = size.container_style();

    html! {
        <>
        <CongratsDialog/>
        <HistoryDialog/>
        <div class="outer-container"  style={outer_container_style}>
            <div class="container"   style={container_style}>
                <Circles  />
                <Crosshairs />
                <InfoBar/>
                <TabHeaders />
                <AllFoundWords {cheat} />

                <RecentWords />

                <canvas id="confetti-canvas" class="confetti-canvas"></canvas>

            </div>
        </div>
        </>
    }
}
