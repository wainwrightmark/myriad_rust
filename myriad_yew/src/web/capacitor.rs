use std::future::Future;

use capacitor_bindings::helpers::Error;

use crate::state::logging;

pub fn do_or_report_error<Fut: Future<Output = Result<(), Error>>, F: Fn() -> Fut + 'static>(f: F) {
    yew::platform::spawn_local(async move { do_or_report_error_async(f).await })
}

pub async fn do_or_report_error_async<
    Fut: Future<Output = Result<(), Error>>,
    F: Fn() -> Fut + 'static,
>(
    f: F,
) {
    let r = f().await;

    match r {
        Ok(_) => {}
        Err(err) => {
            log::error!("{err:?}");
            logging::LoggableEvent::try_log_error_message_async(err.to_string()).await;
        }
    }
}

pub async fn get_or_log_error_async<
    T,
    Fut: Future<Output = Result<T, Error>>,
    F: Fn() -> Fut + 'static,
>(
    f: F,
) -> Option<T> {
    let r = f().await;

    match r {
        Ok(data) => Some(data),
        Err(err) => {
            logging::LoggableEvent::try_log_error_message_async(err.to_string()).await;
            None
        }
    }
}
