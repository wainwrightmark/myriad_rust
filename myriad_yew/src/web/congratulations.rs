use crate::state::{prelude::*, game_rating::{GameRating, SuboptimalWord}};
use chrono::Duration;
use myriad::prelude::{Board, FoundWord};
use yew::prelude::*;
use yew_router::prelude::use_navigator;
use yewdux::prelude::*;



#[function_component(CongratsDialog)]
pub fn congrats_dialog() -> Html {
    let navigator = use_navigator().unwrap();
    let (dialog_state, dispatch) = use_store::<DialogState>();

    let on_ok = dispatch.reduce_mut_callback(|state| state.congratulations_dialog_type = None);
    let on_share = dispatch.reduce_mut_callback(|state| {
        state.congratulations_dialog_type = None;
        crate::web::sharing::share();
    });



    let on_new_game = dispatch.reduce_mut_callback(move |state| {
        state.congratulations_dialog_type = None;
        crate::state::msg::move_to_new_game(false, &navigator);
    });

    let timing = use_selector(|state: &FullGameState| state.timing.clone());

    let rating = use_selector(|state: &FullGameState| GameRating::create(state));
    let board = use_selector(|state: & FullGameState| state.game.board.clone());

    if let Some(dialog_type) = dialog_state.congratulations_dialog_type {
        let message: &str = match dialog_type {
            CongratsDialogType::OneHundred => "Well done, you got ðŸ’¯!",
        };

        let time_box = match *timing {
            GameTiming::Cheat => html!(<>

                <p class="time-display">{"You Cheated!"}</p>
                 </>),
            GameTiming::Started { .. } | GameTiming::Unknown => html!(<></>),
            GameTiming::Finished { total_milliseconds } => {
                let total = Duration::milliseconds(total_milliseconds as i64);
                if total.num_minutes() >= 100 {
                    html!(<></>)
                } else {
                    let minutes = total.num_minutes();
                    let seconds = total.num_seconds() - (60 * minutes);
                    let time_string = format!("{minutes:02}:{seconds:02}");
                    html!(<>
                        <p class="time-display">{time_string}</p>
                         </>)
                }
            }
        };

        let rating_box = rating_box(&rating, &board);

        html!(
            <dialog style="top: 25%" open={true}>
                    <p class="dialog-message">{message}</p>
                    {time_box}
                    {rating_box}
                <div class="dialog-buttons">
                    <button class="dialog-button" onclick={on_ok}>{"Ok"}</button>
                    <button class="dialog-button" onclick={on_share}>{"Share"}</button>
                    <button class="dialog-button" onclick={on_new_game}>{"New Game"}</button>
                </div>
          </dialog>
        )
    } else {
        html!(<></>)
    }
}

fn rating_box(game_rating: &GameRating, board: &Board<GRID_ROWS, GRID_COLUMNS, GRID_SIZE>)->Html{

    let score_box =

    if game_rating.actual_steps == game_rating.min_steps{
        html!(
            <p class="score-display">{"Perfect Score"}</p>
        )
    }
    else{
        html!(
            <p class="score-display">{format!("{}/{}", game_rating.actual_steps, game_rating.min_steps)}</p>
        )
    };


    let suboptimal = if !game_rating.suboptimal_words.is_empty(){
        let ws: Html = game_rating.suboptimal_words.iter().map(|x| suboptimal_word(x, board)).collect();
        html!(
            <>
            <details>
                <summary class="suboptimal-summary">
                    {format!("{} Suboptimal {}", game_rating.suboptimal_words.len(), if game_rating.suboptimal_words.len() == 1 {"Solution"} else {"Solutions"})}
                    <span class="icon">{"â†“"}</span>

                </summary>
                <div style="max-height: 200px; overflow-y:scroll;">
                <table class="suboptimal-table">
                {ws}
                </table>
                </div>
            </details>
            <br/>
            </>
        )
    } else{
        html!(<></>)
    };

    let hard = if !game_rating.hard_words.is_empty(){
        let ws: Html = game_rating.hard_words.iter().map(|x| hard_word(x, board)).collect();
        html!(
            <>
            <details>
                <summary class="hard-summary">
                    {format!("{} Hard {}", game_rating.hard_words.len(), if game_rating.suboptimal_words.len() == 1 {"Solution"} else {"Solutions"})}
                    <span class="icon">{"â†“"}</span>

                </summary>
                {ws}
            </details>
            <br/>
            </>
        )
    } else{
        html!(<></>)
    };


    html!(
        <>
        {score_box}
        {suboptimal}
        {hard}
        </>
    )
}

fn suboptimal_word(w: &SuboptimalWord, board: &Board<GRID_ROWS, GRID_COLUMNS, GRID_SIZE> )-> Html{
    html!(
        <tr>
        <td>{w.result()}</td>
        <td>{w.actual.runes(board)}</td>
        <td>{w.best.runes(board)}</td>

        </tr>

    )
}

fn hard_word(w: &FoundWord<GRID_ROWS, GRID_COLUMNS, GRID_SIZE>, board: &Board<GRID_ROWS, GRID_COLUMNS, GRID_SIZE>)-> Html{
    html!(<p class="hardest-display">{format!("{} = {}", w.runes(board), w.result)}</p>)
}
