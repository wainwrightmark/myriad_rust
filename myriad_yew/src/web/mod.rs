mod app;
mod circle;
mod confetti;
mod control_buttons;
mod crosshair;
mod dialog;
mod found_words;
mod icons;
mod info_bar;
mod recent_words;
mod sharing;
mod tab_headers;
mod startup;
pub mod app_redirect;
pub mod capacitor;
pub mod prelude {

    pub use crate::web::app::*;
    pub use crate::web::circle::*;
    pub use crate::web::confetti::*;
    pub use crate::web::control_buttons::*;
    pub use crate::web::crosshair::*;
    pub use crate::web::dialog::*;
    pub use crate::web::found_words::*;
    pub use crate::web::icons::*;
    pub use crate::web::info_bar::*;
    pub use crate::web::recent_words::*;
    pub use crate::web::tab_headers::*;
    pub use crate::web::startup::*;

    pub fn format_number(num: i32) -> String {
        if num == 100 {
            "ðŸ’¯".to_string()
        } else if num < 10 {
            format!("{num:0>1}")
        } else {
            format!("{num:0>2}")
        }
    }

    pub const CIRCLE_RATIO: f32 = 0.8;

    // pub const SQUARE_SIZE: f32 = 120.0;
    // pub const SQUARE_MIDPOINT: f32 = SQUARE_SIZE / 2.0;
    // pub const SVG_WIDTH: f32 = SQUARE_SIZE * 3.0;
    // pub const SVG_HEIGHT: f32 = BOARD_HEIGHT
    //     + TAB_HEADER_TOP_MARGIN
    //     + TAB_HEADER_HEIGHT
    //     + FOUND_WORD_HEIGHT * 2.0
    //     + FOUND_WORD_MARGIN * 3.0;

    // pub const BOARD_HEIGHT: f32 = SQUARE_SIZE * 3.0;

    // pub const TAB_HEADER_TOP_MARGIN: f32 = 60.0;

    // pub const TAB_HEADER_MARGIN: f32 = 6.0;
    // pub const TAB_HEADER_WIDTH: f32 = 50.0;
    // pub const TAB_HEADER_HEIGHT: f32 = 50.0;
    // pub const TAB_HEADER_PADDING: f32 =
    //     (SVG_WIDTH - (TAB_HEADER_WIDTH * 6.0 + TAB_HEADER_MARGIN * 5.0)) / 2.0;

    // pub const FOUND_WORD_WIDTH: f32 = 30.0;
    // pub const FOUND_WORD_HEIGHT: f32 = 30.0;
    // pub const FOUND_WORD_MARGIN: f32 = 5.0;
    // pub const FOUND_WORD_PADDING: f32 =
    //     (SVG_WIDTH - (FOUND_WORD_WIDTH * 10.0 + FOUND_WORD_MARGIN * 9.0)) / 2.0;
}
