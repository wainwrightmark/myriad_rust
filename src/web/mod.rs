pub mod board;
pub mod foundwords;
pub mod newgamebutton;
pub mod recentwords;
pub mod confetti;
