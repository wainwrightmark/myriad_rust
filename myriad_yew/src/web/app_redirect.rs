#[cfg(any(feature = "android", feature = "ios"))]
pub async fn subscribe_to_app_url_events() {
    use capacitor_bindings::app::App;
    //use crate::state::logging::LoggableEvent;
    // LoggableEvent::try_log(LoggableEvent::Internal { message: format!("Subscribing to url open events") });
    if let Ok(handle) = App::add_app_url_open_listener(|x| redirect_to_url(x.url)).await {
        handle.leak();
    }    
}

#[cfg(any(feature = "android", feature = "ios"))]
fn redirect_to_url(url: String) {    
    use web_sys::window;

    //use crate::state::logging::LoggableEvent;
    // LoggableEvent::try_log(LoggableEvent::Internal { message: format!("Redirect to url: {url}") });
    let Some(url)  = web_sys::Url::new(&url).ok() else {
        return ;};
    let Some(window) = window() else{return ;};

    let Ok(protocol) = window.location().protocol()else {
        return ;};
    let Ok(host) = window.location().host()else {

        return ;};

    url.set_protocol(&protocol);
    url.set_host(&host);
    let url = url.href();
    let Ok(current_href) = window.location().href() else {return ;};

    if current_href != url {
        let _ = window.location().set_href(&url);
        //let _ = window.history().unwrap().push_state_with_url(&wasm_bindgen::JsValue::NULL, "", Some(&url));
    }
}
