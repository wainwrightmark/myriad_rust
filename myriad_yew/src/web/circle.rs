use std::ops::Deref;

use crate::state::prelude::*;
use crate::web::prelude::*;
use myriad::prelude::*;
use wasm_bindgen::JsCast;
use web_sys::*;
use yew::prelude::*;
use yewdux::prelude::*;

#[function_component(Circles)]
pub fn circles() -> Html {
    let circles = Tile::<GRID_COLUMNS, GRID_ROWS>::iter_by_row()
        .map(|coordinate| html!(< Circle {coordinate} />))
        .collect::<Html>();

    let ontouchmove = Dispatch::new().apply_callback(move |ev: TouchEvent| {
        if let Some(tile) = get_tile_from_touch_event(ev) {
            InputMsg::Enter { coordinate: tile }
        } else {
            InputMsg::None
        }
    });

    let onpointerup = Dispatch::new().apply_callback(move |_: PointerEvent| InputMsg::Up {});

    html! {
      <div id="circles" class="circles" {onpointerup} {ontouchmove}>
          {circles}
      </div>

    }
}

fn get_tile_from_touch_event(ev: TouchEvent) -> Option<Tile<GRID_COLUMNS, GRID_ROWS>> {
    let touch = ev.target_touches().item(0)?;

    let window = window()?;
    let document = window.document()?;
    let x = touch.page_x();
    let y = touch.page_y();
    let element = document.element_from_point(x as f32, y as f32)?;

    let id = element.id();

    //log::info!("touch move to {id}");

    TILES.get(id.as_str()).cloned()
}

use phf::phf_map;

static TILES: phf::Map<&'static str, Tile<GRID_COLUMNS, GRID_ROWS>> = phf_map! {
    "(0,0)_bigCircle" => Tile::new_const::<0,0>(),
    "(0,1)_bigCircle" => Tile::new_const::<0,1>(),
    "(0,2)_bigCircle" => Tile::new_const::<0,2>(),
    "(1,0)_bigCircle" => Tile::new_const::<1,0>(),
    "(1,1)_bigCircle" => Tile::new_const::<1,1>(),
    "(1,2)_bigCircle" => Tile::new_const::<1,2>(),
    "(2,0)_bigCircle" => Tile::new_const::<2,0>(),
    "(2,1)_bigCircle" => Tile::new_const::<2,1>(),
    "(2,2)_bigCircle" => Tile::new_const::<2,2>(),
};

#[derive(PartialEq, Properties)]
pub struct CircleProperties {
    pub coordinate: Tile<GRID_COLUMNS, GRID_ROWS>,
}
#[function_component(Circle)]
fn circle(properties: &CircleProperties) -> Html {
    let coordinate = properties.coordinate;

    let game_size = *use_store::<GameSize>().0.as_ref();
    let location = use_selector_with_deps(
        |state: &RotFlipState, (co, size)| state.get_location(co, size),
        (coordinate, game_size),
    );

    let board = use_selector(|state: &FullGameState| state.game.board.clone());

    let circle_type = *use_selector_with_deps(
        |state: &ChosenPositionsState, (co, board)| state.get_circle_type(co, board),
        (coordinate, board),
    )
    .deref();

    let letter = *use_selector_with_deps(
        |state: &FullGameState, co| state.game.board[*co],
        coordinate,
    )
    .deref();

    let onpointerdown = Dispatch::new().apply_callback(move |ev: PointerEvent| {
        ev.target()
            .unwrap()
            .dyn_into::<Element>()
            .unwrap()
            .release_pointer_capture(ev.pointer_id())
            .unwrap();
        InputMsg::Down { coordinate }
    });

    let onpointerenter =
        Dispatch::new().apply_callback(move |_: PointerEvent| InputMsg::Enter { coordinate });

    let onpointerup = Dispatch::new().apply_callback(move |_: PointerEvent| InputMsg::Up {});

    let square_radius = game_size.square_radius();

    let left = location.x - (square_radius * CIRCLE_RATIO);
    let top = location.y - (square_radius * CIRCLE_RATIO);

    let text = if matches!(letter, Rune::Blank) {
        "".to_string()
    } else {
        letter.to_string()
    };
    let key = format!("{coordinate}_key");
    let circle_id = format!("{coordinate}_bigCircle");
    let text_id = format!("{coordinate}_text");

    let g_style = format!("left: {left}px; top: {top}px;");

    let circle_type_class = match circle_type {
        CircleType::Disabled => "circle-disabled",
        CircleType::LegalMove => "circle-legal",
        CircleType::LastPosition => "circle-final",
        CircleType::IntermediatePosition { next: _ } => "circle-intermediate",
    };

    let circle_classes = classes!("circle", circle_type_class);

    html! {
        <div class="square" {key} style={g_style}>

            <div id={circle_id} class={circle_classes} {onpointerdown} {onpointerup} {onpointerenter}>
                <p id={text_id} class="circle-text"> {text} </p>
            </div>
        </div>
    }
}
