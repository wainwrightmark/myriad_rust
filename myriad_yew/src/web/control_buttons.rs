use crate::state::full_game_state::FullGameState;
use crate::state::game_size::*;
use crate::state::msg;
use crate::state::prelude::*;
use crate::web::prelude::*;
use yew::prelude::*;
use yewdux::prelude::*;

#[derive(PartialEq, Properties)]
pub struct GameButtonProperties {
    pub selected_tab: usize,

    pub position_number: i32,
    pub width: f32,
}

#[function_component(TodayGameButton)]
pub fn todays_game_button(properties: &GameButtonProperties) -> Html {
    let (game_size, _) = use_store::<GameSize>();
    let on_click: Option<Callback<MouseEvent>> = Some(Callback::<MouseEvent>::from(|_| {msg::move_to_new_game(true);}));
        //Some(Dispatch::new().apply_callback(|_| NewGameMsg { today: true }));

    let (x, y) = game_size.get_found_word_position(
        properties.position_number,
        properties.selected_tab,
        false,
    );

    html!(<ButtonBox id={"today_game_button"} text={"Today's Game"} {x} {y} width_units={properties.width} color="white" {on_click} />)
}

#[function_component(RandomGameButton)]
pub fn random_game_button(properties: &GameButtonProperties) -> Html {
    let (game_size, _) = use_store::<GameSize>();
    let on_click: Option<Callback<MouseEvent>> = Some(Callback::<MouseEvent>::from(|_| {msg::move_to_new_game(false);}));

    let (x, y) = game_size.get_found_word_position(
        properties.position_number,
        properties.selected_tab,
        false,
    );

    html!(<ButtonBox id={"random_game_button"} text={"Random Game"} {x} {y} width_units={properties.width} color="white" {on_click} />)
}

#[derive(PartialEq, Properties)]
pub struct ScoreCounterProperties {
    pub total_found: usize,
    pub selected_tab: usize,

    pub position_number: i32,
    pub width: f32,
}

#[function_component(ScoreCounter)]
pub fn score_counter(properties: &ScoreCounterProperties) -> Html {
    let (game_size, _) = use_store::<GameSize>();
    let (x, y) = game_size.get_found_word_position(
        properties.position_number,
        properties.selected_tab,
        false,
    );

    let (found,total) = *use_selector(|state: &FullGameState| state.get_found_count());
    let found_pc = found * 100 / total;
    let gradient = format!("background: linear-gradient(to right, green {found_pc}%, lightgrey {found_pc}%, lightgrey);");

    let width = format!("{}", FOUND_WORD_WIDTH * properties.width);
    let height = format!("{FOUND_WORD_HEIGHT}");
    let style = format!("position:absolute; transform: translate({x}px, {y}px); height: {height}px; width: {width}px; border-radius:5px; {gradient}");

    let class = classes!("found-word",);
    let key = "score_counter";

    let text = format_number(properties.total_found as i32);

    html!(
        <button {key} {style} {class} >
            {text}
        </button>
    )
}

#[function_component(RotateButton)]
pub fn rotate_button(properties: &GameButtonProperties) -> Html {
    let (game_size, _) = use_store::<GameSize>();
    let on_click: Option<Callback<MouseEvent>> =
        Some(Dispatch::new().apply_callback(|_| RotFlipMsg {
            rotate: myriad::prelude::QuarterTurns::One,
            flip: false,
        }));

    let (x, y) = game_size.get_found_word_position(
        properties.position_number,
        properties.selected_tab,
        false,
    );

    html!(<ButtonBox id={"rotate_button"} text={"‚ü≥"} {x} {y} width_units={properties.width} color="white" {on_click} />)
}

#[function_component(FlipButton)]
pub fn flip_button(properties: &GameButtonProperties) -> Html {
    let (game_size, _) = use_store::<GameSize>();
    let on_click: Option<Callback<MouseEvent>> =
        Some(Dispatch::new().apply_callback(|_| RotFlipMsg {
            rotate: myriad::prelude::QuarterTurns::Zero,
            flip: true,
        }));

    let (x, y) = game_size.get_found_word_position(
        properties.position_number,
        properties.selected_tab,
        false,
    );

    html!(<ButtonBox id={"flip_button"} text={"‚¨å"} {x} {y} width_units={properties.width} color="white" {on_click} />)
}

#[function_component(HistoryButton)]
pub fn history_button(properties: &GameButtonProperties) -> Html {
    let (game_size, _) = use_store::<GameSize>();
    let on_click: Option<Callback<MouseEvent>> = Some(
        Dispatch::<DialogState>::new()
            .reduce_mut_callback(|s| s.history_dialog_type = Some(Default::default())),
    );

    let (x, y) = game_size.get_found_word_position(
        properties.position_number,
        properties.selected_tab,
        false,
    );

    html!(<ButtonBox id={"history_button"} text={"H"} {x} {y} width_units={properties.width} color="white" {on_click} />)
}

// #[function_component(WainwrongButton)]
// pub fn wainwrong_button(properties: &GameButtonProperties) -> Html {
//     let (x, y) =properties.game_size.
//         get_found_word_position(properties.position_number, properties.selected_tab, false);

//     let rect_class = classes!("found-word-box", "found-word-box-button");
//     let text_class = classes!("button-text");
//     let class = classes!("found-word", "found-word-button");

//     let style = format!("transform: translate({x}px, {y}px);");

//     html!(
//      <g key={"wainwrong_link"} {style} {class} >

//      <a href="https://wainwrong.com/">
//      <rect class={rect_class} height={format!("{FOUND_WORD_HEIGHT}")} rx="5" width={format!("{}", FOUND_WORD_WIDTH * properties.width)}>

//      </rect>
//      <text class={text_class}>
//      {"ùï®"}
//      </text>

//      </a>
//      </g>)
// }

// #[function_component(FacebookButton)]
// pub fn facebook_button(properties: &GameButtonProperties) -> Html {
//     let (x, y) =properties.game_size.
//         get_found_word_position(properties.position_number, properties.selected_tab, false);

//     let rect_class = classes!("found-word-box", "found-word-box-button");
//     let class = classes!("found-word", "found-word-button");

//     let style = format!("transform: translate({x}px, {y}px);");

//     html!(
//         <g key={"share_link"} {style} {class} >

//         <a href="https://www.facebook.com/sharer/sharer.php?u=https%3A%2F%2Fwainwrightmark.github.io%2Fmyriad_rust">
//         <rect class={rect_class} height={format!("{FOUND_WORD_HEIGHT}")} rx="5" width={format!("{}", FOUND_WORD_WIDTH * properties.width)}>

//         </rect>

//         <svg data-license="From https://github.com/twbs/icons - Licensed under MIT"
//         fill="currentColor"
//         height="24"
//         style="margin: 0.1em; display: initial;"
//         viewBox="0 0 16 16"
//         width="24"
//         xmlns="http://www.w3.org/2000/svg">
//         <title>{"BootstrapFacebook"}</title>
//         <path d={FACEBOOK_ICON_PATH}></path></svg>

//         </a>
//         </g>)
// }

#[derive(PartialEq, Properties)]
pub struct ButtonBoxProperties {
    pub id: String,
    pub text: AttrValue,
    pub color: AttrValue,
    pub x: f32,
    pub y: f32,
    pub width_units: f32,
    pub on_click: Option<Callback<MouseEvent>>,
}

#[function_component(ButtonBox)]
fn button_box(properties: &ButtonBoxProperties) -> Html {
    let x = properties.x;
    let y = properties.y;
    let width = format!("{}", FOUND_WORD_WIDTH * properties.width_units);
    let height = format!("{FOUND_WORD_HEIGHT}");
    let color = &properties.color;
    let style = format!("position:absolute; transform: translate({x}px, {y}px); height: {height}px; width: {width}px; border-radius:5px; background-color: {color};");

    let class = classes!(
        "found-word",
        if properties.on_click.is_some() {
            Some("found-word-button")
        } else {
            None
        }
    );
    let key = properties.id.clone();

    html!(
        <button {key} {style} {class} onclick={properties.on_click.clone()}>
            {properties.text.clone()}
        </button>
    )
}
