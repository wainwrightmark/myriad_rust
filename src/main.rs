use yew::prelude::*;

use crate::web::prelude::*;
pub mod core;
pub mod state;
pub mod web;

#[function_component(App)]
fn app() -> Html {
    html! {
        <div class="container">

        <NewGameButton/>
        <svg viewBox="0 0 120 120" class="myriadSVG">
        <rect x="0" y="0" width="120" height="120" fill="white"/>
        <RopeSVG />
        <CirclesSVG />
        
        <RecentWords/>
        </svg>
        <FoundWordsTable/>
        

        <canvas id="confetti-canvas" style="position: fixed; width: 100%; height: 100%; top: 0px; left: 0px; z-index: 1000; pointer-events: none;"></canvas>

        </div>


    }
}

fn main() {
    wasm_logger::init(wasm_logger::Config::default());
    yew::Renderer::<App>::new().render();
}
