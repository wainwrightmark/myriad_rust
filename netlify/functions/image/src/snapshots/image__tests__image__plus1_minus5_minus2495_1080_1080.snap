---
source: src/main.rs
expression: hash
---
4557080504673849569
